File: 687-longest-univalue-path/solution.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

class BTreeNode {
  val: number;
  left: BTreeNode | null;
  right: BTreeNode | null;
  constructor(val?: number, left?: BTreeNode | null, right?: BTreeNode | null) {
    this.val = val === undefined ? 0 : val;
    this.left = left === undefined ? null : left;
    this.right = right === undefined ? null : right;
  }
}

// https://leetcode.com/problems/longest-univalue-path/solutions/3045413/simple-dfs-solution-beats-92-o-n-time:
function dfs(node: BTreeNode | null, answer: number[]): number {
  if (!node) return answer[0];

  let left = dfs(node.left, answer);
  let right = dfs(node.right, answer);

  if (node.left?.val !== node.val) left = 0;
  if (node.right?.val !== node.val) right = 0;
  answer[0] = Math.max(answer[0], left + right);

  return 1 + Math.max(left, right);
}

function longestUnivaluePath(root: BTreeNode | null): number {
  const answer = [0];
  dfs(root, answer);
  return answer[0];
}

File: 687-longest-univalue-path/solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int traverse(int& ans, TreeNode* node) {
        int len = 1, size = 1, sl = 0, sr = 0, val = node->val;
        if(node->left != NULL) {
            int vl = node->left->val;
            sl = traverse(ans, node->left);
            if(val == vl) {
                len += sl;
                size = max(size, 1 + sl);
            }
        }
        if(node->right != NULL) {
            int vr = node->right->val;
            int sr = traverse(ans, node->right);
            if(val == vr) {
                len += sr;
                size = max(size, 1 + sr);
            }
        }
        ans = max(ans, len);
        return size;
    }

    int longestUnivaluePath(TreeNode* root) {
        if(root == NULL) return 0;
        int ans = 0;
        traverse(ans, root);
        return max(ans - 1, 0);
    }
};
